### ISSUE TITLE
`comm_subplan_elim` segfault on main `StackOverflow`

### ISSUE THREAD

**Issue by cmdlineluser** [pola-rs/polars#15345]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

lf = pl.LazyFrame({"A": []})

lf = pl.concat([lf] * 8192)

lf.with_columns(1).collect(comm_subplan_elim=True) # default, but works with False 
# zsh: segmentation fault
```

### Log output

```shell
found multiple sources; run comm_subplan_elim
```

### Issue description

I guess this is related to the new CSE work.

It only seems to trigger for `>= 8192`

### Expected behavior

No segfault.

### Installed versions

<details>

```
main
```

</details>

**Comment by mcrumiller** [comment]
I cannot reproduce on latest main:

```python
import polars as pl
lf = pl.LazyFrame({"A": []})
lf = pl.concat([lf] * 8192)
lf.with_columns(1).collect(comm_subplan_elim=True)
# shape: (0, 2)
# ┌──────┬─────────┐
# │ A    ┆ literal │
# │ ---  ┆ ---     │
# │ null ┆ i32     │
# ╞══════╪═════════╡
# └──────┴─────────┘
```

**Comment by ritchie46** [comment]
This is a stack overflow. We hope to be able fix this properly by being able to set the stack on the heap if it grows too big, but it isn't easy.

**Comment by ritchie46** [comment]
mostly fixed by #15355
